docs/sampling.rst

```diff
@@ -0,0 +1,133 @@
+still too expensive, because it forces us to still throw a random number from
+the Poisson distribution (which is expensive) per allocation. What we do instead
+(and other sampling profilers as well) is to use the Exponential distribution.
+The exponential distribution tells us the amount of bytes we need to wait until
+the next sample needs to be taken. This is much better than throwing a random
+number per allocation, because in this way we need to throw one random number
+(that gives us how many bytes we need to wait for), then keep substracting bytes
+to this number until it reaches 0 and then we can take a sample and do it all
+over again. This means that for many allocations we just need to substract from
+the number of remaining bytes instead of throwing a random number.
+
+As is customary in this method, an optimization here is that if a given
+allocation is bigger than the average ammount of bytes we need to wait until the
+next sample, we take it directly as is very likely that we are going to sample
+it at least once.
+
+All this process is dependent on the probability of sampling a single byte in
+the Bernoulli process, which is the inverse of how many bytes on average we need
+to wait until the next sample. Choosing this number wisely is very important for
+good results.
+
+.. note:: 
+
+    This random distribution approach has several advantages: first, its
+    statistical properties make then very easy to estimate actual statistical
+    characteristics of the heap. Indeed, we certainly want large allocated
+    blocks to weight more than small ones. Second, Poisson process are
+    relatively easy to simulate with reasonable performance.
+
+Algorithm
+---------
+
+As many other profilers that implement sampling using this method, the algorith `memray`
+follows is:
+
+* We draw a random number from the exponential distribution with a given sample rate. This tells us how many bytes
+  we must wait until taking a sample.
+
+* An allocation is being tracked.
+
+* If the allocation size is bigger than the sampling size, we sample the entire allocation with its original size. We do
+  this becuse the probability tht we are going to take at least one sample is very high and therefore is a good performance
+  optimization.
+
+* If the allocation size smaller than our sampling size, then we substract the size of the allocation from the ammount of bytes
+  we need to wait until we take a sample. If this number reaches zero, we take a sample and we draw another random number from
+  the exponential distribution. The amount of times we had to draw from the exponential distribution is the number of samples
+  the allocation should count as.
+
+* We multiply the number of samples we drew within the allocation by the sampling rate to get an estimate of the statistical
+  size of the allocation. We record this allocation and its stack as if the allocation had the statistical size that we obtained.
```

Comment:
Not sure I completely follow, but if you are proposing with "ust track the exact number of bytes allocated since our last sample, and weight the new sample as that size" to just take a sample every nth byte, then the problem is that you are biasing towards the allocation pattern of the application. If the application has cyclic/periodic/regular allocation patterns there is a lot of them that you will never see, that's why you need an unbiased statistical approach.